test/setup/TestEnv.ts

```diff
@@ -0,0 +1,137 @@
+import { mkdirSync, rmSync } from 'fs';
+import { spawnSync } from 'child_process';
+import { SevereServiceError } from 'webdriverio';
+import TestSettings from '../types/TestSettings.js';
+import checkIfUp from './checkIfUp.js';
+import logger, { Logger } from '@wdio/logger';
+import { makeTestSettings } from './makeTestSettings.js';
+
+declare global {
+	// eslint-disable-next-line no-var, no-use-before-define
+	var testEnv: TestEnv | undefined;
+}
+
+export default class TestEnv {
+	public settings: TestSettings;
+	public testLog: Logger;
+	public baseDockerComposeCmd: string;
+
+	public static createWithDefaults(
+		providedSettings: Partial<TestSettings>
+	): TestEnv {
+		const settings = makeTestSettings( providedSettings );
+		return new this( settings );
+	}
+
+	public constructor( settings?: TestSettings ) {
+		if ( !settings ) {
+			throw new Error(
+				'Settings are required to create a Test Environment'
+			);
+		}
+
+		this.settings = settings;
+		this.testLog = logger( 'test-env' );
+		this.testLog.setDefaultLevel( 'debug' );
+		this.baseDockerComposeCmd = this.makeBaseDockerComposeCmd();
+	}
+
+	public get vars(): Record<string, string> {
+		return this.settings.vars;
+	}
+
+	public async up(): Promise<void> {
+		try {
+			process.once( 'SIGINT', async () => {
+				await this.down();
+				// eslint-disable-next-line no-process-exit
+				process.exit( 1 );
+			} );
+
+			this.resetOutputDir();
+			await this.settings.beforeServices();
+
+			console.log( '▶️  Bringing up the test environment' );
+
+			this.stopServices();
+			this.startServices();
+			await this.waitForServices();
+
+			if ( this.settings.runHeaded ) {
+				console.log(
+					'💻 Open http://localhost:7900/?autoconnect=1&resize=scale to observe headed tests.\n'
+				);
+			}
+		} catch ( e ) {
+			throw new SevereServiceError( e );
+		}
+	}
+
+	public async down(): Promise<void> {
+		this.stopServices();
+	}
+
+	public async waitForServices(): Promise<void[]> {
+		return Promise.all( this.settings.waitForUrls().map(
+			async ( waitForURL: string ): Promise<void> => {
+				await checkIfUp( waitForURL, this.settings.testTimeout );
+				this.testLog.info( `ℹ️  Successfully loaded ${waitForURL}` );
+			}
+		) );
+	}
+
+	public runDockerComposeCmd( dockerComposeOptionsCommandAndArgs: string ): string {
+		const dockerComposeCmd = `${this.baseDockerComposeCmd} ${dockerComposeOptionsCommandAndArgs}`;
+
+		this.testLog.debug( 'Running: ', dockerComposeCmd );
+
+		const result = spawnSync( dockerComposeCmd, {
+			stdio: 'pipe',
+			shell: true,
+			encoding: 'utf-8',
+			env: { ...this.vars, OUTPUT_DIR: this.settings.outputDir, PATH: process.env.PATH }
+		} );
```

Comment:
Just curious - what's the thinking behind making this `sync` instead of `async`?

Reply:
Has the knock-on effect of making `dbQuery` `sync`, and I'm wondering what the advantages or disadvantages of that is

Reply:
I think my previous explanation breezed over this a bit... I share your ambivalence, and am open to switching to the `child_process.sync`, but don't prefer it for these reasons:

Reason 1: In `TestEnv` I'm only spawning Docker commands and in all cases we do always want to wait for their completion before executing anything else in the Node side. This stuff is happening exactly at the boundary between our "test runner" / test code and the "test services" we are testing and managing through Docker. `TestEnv`'s job is largely to handle all the things that are not about running the test strictly speaking, but are about setting up the environment in which the tests runs, which is largely a Docker world in our setup. I think you get all of that, but lay it out for context. So using `spawnSync` is helpful here as it ensures that we block Node execution at each of these steps when we otherwise, as a rule, want to make sure we do that. This is probably true with most of our async functions as well, but I don't mind and kinda like it being enforced and highlighted in the Docker / `TestEnv` stuff here exactly because it is a crossing of worlds. 

Reason 2: It is a little unexpected but Node's `child_process.spawn` (vs `spawnSync`) is async but isn't implemented to the async/await pattern. The API is not particular ergonomic and would likely warrant another helper function to re-package `child_process.spawn` in an async/await way, or use one of a few commonly used npm modules that would to that for us. However, pragmatically the effect would be the same in our use. We would alway want to await the Docker command to fully complete before proceeding executing any test code. That said, if Node's async `spawn` function was implemented to the async/await interface already I'd probably just have used that to make it look like everything else.

Reply:
Here is what I've done: I've went ahead and converted `runDockerComposeCmd` to be async while still using `spawnSync` internally. This keeps `browser.dbQuery` async like all the other `browser` commands as well as makes the private methods `startServices` and `stopServices`  on `TestEnv` also async. This should make things more consistent with the rest of our code while not unnecessarily getting into using an async `spawn` setup.